scripts/merge-script/index.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as Octokit from '@octokit/rest';
+import {spawnSync, SpawnSyncOptions, SpawnSyncReturns} from 'child_process';
+import {join} from 'path';
+import {Config} from './config';
+import {GitClient, GitCommandError} from './git';
+import {
+  isPullRequest,
+  loadAndValidatePullRequest,
+  PullRequest,
+  PullRequestFailure,
+  PullRequestStatus
+} from './pull-request';
+import {GithubApiMergeStrategy} from './strategies/api-merge';
+import {AutosquashMergeStrategy} from './strategies/autosquash-merge';
+import {matchesPattern} from './string-pattern';
+
+/** Describes the status of a pull request merge. */
+export const enum MergeStatus {
+  UNKNOWN_GIT_ERROR,
+  DIRTY_WORKING_DIR,
+  SUCCESS,
+  FAILED,
+}
+
+/** Result of a pull request merge. */
+export interface MergeResult {
+  /** Overall status of the merge. */
+  status: MergeStatus;
+  /** List of pull request failures. */
+  failure?: PullRequestFailure;
+}
+
+
+/**
+ * Class that accepts a merge script configuration and Github token. It provides
+ * a programmatic interface for merging multiple pull requests based on their
+ * labels that have been resolved through the merge script configuration.
+ */
+export class PullRequestMergeTask {
+  /** Git client that can be used to execute Git commands. */
+  git = new GitClient(this._githubToken, this.config);
+
+  constructor(public config: Config, private _githubToken: string) {}
+
+  /**
+   * Merges the given pull request and pushes it upstream.
+   * @param prNumber Pull request that should be merged.
+   * @param force Whether non-critical pull request failures should be ignored.
+   */
+  async merge(prNumber: number, force = false): Promise<MergeResult> {
+    if (this._hasUncommittedChanges()) {
+      return {status: MergeStatus.DIRTY_WORKING_DIR};
+    }
+
+    const pullRequestOrFailure = await loadAndValidatePullRequest(this, prNumber, force);
+
+    if (!isPullRequest(pullRequestOrFailure)) {
+      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
+    }
+
+    const pullRequest: PullRequest = pullRequestOrFailure;
```

Comment:
Is it possible for this to be done with a type guard?

`isPullRequest` can confirm the type of `pullRequestOrFailure`?

That would allow this to be:

```suggestion
    const pullRequest = await loadAndValidatePullRequest(this, prNumber, force);

    if (!isPullRequest(pullRequest)) {
      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
    }
```

Reply:
`isPullRequest` already is a type guard. The only difference I see here is that you renamed the variable from `pullRequestOrFailure` to `pullRequest`. I'm fine with that.